netket/optimizer/jax_stochastic_reconfiguration.py

```diff
@@ -0,0 +1,370 @@
+
+                self._x0 = out
+            else:
+                self._S = _np.matmul(oks.conj().T, oks, self._S)
+                self._S = _sum_inplace(self._S)
+                self._S /= float(n_samp)
+
+                self._apply_preconditioning(grad)
+
+                if self._lsq_solver == "Cholesky":
+                    c, low = _cho_factor(self._S, check_finite=False)
+                    out[:] = _cho_solve((c, low), grad)
+
+                else:
+                    out[:], residuals, self._last_rank, s_vals = _lstsq(
+                        self._S,
+                        grad,
+                        cond=self._svd_threshold,
+                        lapack_driver=self._lapack_driver,
+                    )
+
+                self._revert_preconditioning(out)
+
+        else:
+            if self._use_iterative:
+                if self._lsq_solver == "jaxcg":
+                    out.real = self._jax_cg_solve(oks,grad.real,n_samp)
+                else:
+                    op = self._linear_operator(oks, n_samp)
+
+                    if self._x0 is None:
+                        self._x0 = _np.zeros(n_par)
+
+                    out[:].real, info = self._sparse_solver(
+                        op,
+                        grad.real,
+                        x0=self._x0,
+                        tol=self.sparse_tol,
+                        maxiter=self.sparse_maxiter,
+                    )
+                if info < 0:
+                    raise RuntimeError("SR sparse solver did not converge.")
+                
+                self._x0 = out.real
+            else:
+                self._S = _np.matmul(oks.conj().T, oks, self._S)
+                self._S /= float(n_samp)
+
+                self._apply_preconditioning(grad)
+
+                if self._lsq_solver == "Cholesky":
+                    c, low = _cho_factor(self._S, check_finite=False)
+                    out[:].real = _cho_solve((c, low), grad)
+                else:
+                    out[:].real, residuals, self._last_rank, s_vals = _lstsq(
+                        self._S.real,
+                        grad.real,
+                        cond=self._svd_threshold,
+                        lapack_driver=self._lapack_driver,
+                    )
+
+                self._revert_preconditioning(out.real)
+
+            out.imag.fill(0.0)
+        self._comm.bcast(out, root=0)
+        self._comm.barrier()
+        return out
+
+    def _jax_cg_solve(self,oks,grad,n_samp):
+        """
+        Solves the SR flow equation using the conjugate gradient method 
+        """
+
+        n_par = grad.shape[0]
+        if self._x0 is None:
+            self._x0 = _np.zeros(n_par, dtype=_np.complex128)
+
+        cov_op = jax.jit(self._jax_linear_function(oks,n_samp))
+        
+        out, _ = jcg(cov_op,grad,x0=self._x0,tol=self.sparse_tol,maxiter=self.sparse_maxiter)
+
+        return out
+
+    def _jax_linear_function(self,oks,n_samp):
```

Comment:
To avoid recompiling at each call of the SR, I think that the jit decorator should be put here, and not when definining cov_op above. 

Basically 

```python
@partial(jax.jit, static_argnums=(0, 2))
def _jax_linear_function(self,oks,n_samp)
```

Reply:
_jax_linear_function returns the function matvec(cov,x) so you can't jit it. I tried jitting the matvec function inside and it's still slower (also no difference between step 1 and 2 so I think its compiling every time). 

I'll keep thinking about it

Reply:
There is also this thing: 

https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.custom_linear_solve.html#jax.lax.custom_linear_solve

It looks like this leverages jax's autodiff capabilities.

Reply:
Also here's the issue where they're discussing the linear solvers
https://github.com/google/jax/issues/1531

Reply:
but we don't really need a differentiable solver, we just need a solver that works with jax types (thus possibly fast on a gpu) without converting to numpy! In any case we can leave the performance issue aside for the moment, we can fix this later